src/core/rbtree.rs

```diff
@@ -0,0 +1,358 @@
+//! Wrapper over the `ngx_rbtree_t`.
+
+use core::alloc::Layout;
+use core::cmp::Ordering;
+use core::hash::{self, BuildHasher, Hash};
+use core::marker::PhantomData;
+use core::ptr::{self, NonNull};
+use core::{borrow, mem};
+
+use nginx_sys::{
+    ngx_rbt_red, ngx_rbtree_data, ngx_rbtree_delete, ngx_rbtree_init, ngx_rbtree_insert,
+    ngx_rbtree_key_t, ngx_rbtree_min, ngx_rbtree_next, ngx_rbtree_node_t, ngx_rbtree_t,
+};
+
+use crate::allocator::{self, AllocError, Allocator};
+
+#[allow(deprecated)]
+type RbTreeHasher = core::hash::BuildHasherDefault<hash::SipHasher>;
+
+/// Wrapper over the `ngx_rbtree_t`.
+#[derive(Debug)]
+pub struct RbTree<K, V, A>
+where
+    A: Allocator,
+{
+    tree: ngx_rbtree_t,
+    sentinel: NonNull<ngx_rbtree_node_t>,
+    alloc: A,
+    _ph: PhantomData<(K, V)>,
+}
+
+struct Node<K, V> {
+    node: ngx_rbtree_node_t,
+    key: K,
+    value: V,
+}
+
+impl<K, V> Node<K, V>
+where
+    K: Hash,
+{
+    fn new(key: K, value: V) -> Self {
+        let mut node: ngx_rbtree_node_t = unsafe { mem::zeroed() };
+        node.key = RbTreeHasher::default().hash_one(&key) as ngx_rbtree_key_t;
+
+        Self { node, key, value }
+    }
+
+    fn into_kv(self) -> (K, V) {
+        (self.key, self.value)
+    }
+}
+
+/// Raw iterator over the `ngx_rbtree_t` nodes.
+///
+/// This iterator remains valid after removing the current item from the tree.
+pub struct RawIter<'a> {
+    tree: NonNull<ngx_rbtree_t>,
+    node: *mut ngx_rbtree_node_t,
+    _ph: PhantomData<&'a ()>,
+}
+
+impl<'a> RawIter<'a> {
+    /// Creates an iterator for the `ngx_rbtree_t`.
+    pub fn new(tree: NonNull<ngx_rbtree_t>) -> Self {
+        let t = unsafe { tree.as_ref() };
+        let node = if ptr::addr_eq(t.root, t.sentinel) {
+            // empty tree
+            ptr::null_mut()
+        } else {
+            unsafe { ngx_rbtree_min(t.root, t.sentinel) }
+        };
+
+        Self {
+            tree,
+            node,
+            _ph: Default::default(),
```

Comment:
Arbitrary, but I think its more idiomatic to use `PhantomData` here instead of `Default::default()`